src/brod_consumer.erl

```diff
@@ -570,49 +586,59 @@ send_fetch_request(#state{ begin_offset = BeginOffset
-      State4 = State3#state{is_suspended = false},
-      State  = maybe_send_fetch_request(State4),
+      State  = maybe_send_fetch_request(State2),
       {reply, ok, State};
     {error, Reason} ->
       {reply, {error, Reason}, State0}
   end.
 
 %% @private
 -spec update_options(options(), state()) -> {ok, state()} | {error, any()}.
-update_options(Options, #state{begin_offset = OldBeginOffset} = State) ->
+update_options(Options, #state{begin_offset = OldBeginOffset} = State0) ->
   F = fun(Name, Default) -> proplists:get_value(Name, Options, Default) end,
   NewBeginOffset = F(begin_offset, OldBeginOffset),
-  OffsetResetPolicy = F(offset_reset_policy, State#state.offset_reset_policy),
-  State1 = State#state
+  OffsetResetPolicy = F(offset_reset_policy, State0#state.offset_reset_policy),
+  State1 = State0#state
     { begin_offset        = NewBeginOffset
-    , min_bytes           = F(min_bytes, State#state.min_bytes)
-    , max_bytes_orig      = F(max_bytes, State#state.max_bytes_orig)
-    , max_wait_time       = F(max_wait_time, State#state.max_wait_time)
-    , sleep_timeout       = F(sleep_timeout, State#state.sleep_timeout)
-    , prefetch_count      = F(prefetch_count, State#state.prefetch_count)
+    , min_bytes           = F(min_bytes, State0#state.min_bytes)
+    , max_bytes_orig      = F(max_bytes, State0#state.max_bytes_orig)
+    , max_wait_time       = F(max_wait_time, State0#state.max_wait_time)
+    , sleep_timeout       = F(sleep_timeout, State0#state.sleep_timeout)
+    , prefetch_count      = F(prefetch_count, State0#state.prefetch_count)
     , offset_reset_policy = OffsetResetPolicy
-    , max_bytes           = F(max_bytes, State#state.max_bytes)
-    , size_stat_window    = F(size_stat_window, State#state.size_stat_window)
+    , max_bytes           = F(max_bytes, State0#state.max_bytes)
+    , size_stat_window    = F(size_stat_window, State0#state.size_stat_window)
     },
-  NewState =
-    case NewBeginOffset =/= OldBeginOffset of
-      true ->
-        %% reset buffer in case subscriber wants to fetch from a new offset
-        State1#state{pending_acks = #pending_acks{}};
-      false ->
-        State1
-    end,
-  resolve_begin_offset(NewState).
+  case resolve_begin_offset(State1) of
+    {ok, #state{begin_offset = ResolvedBeginOffset} = State2} ->
+      AckedOffset = F(acked_offset, ResolvedBeginOffset - 1),
+      PendingAcks = make_pending_acks(AckedOffset, ResolvedBeginOffset),
+      {ok, State2#state{pending_acks = PendingAcks}};
+    {error, _Reason} = Error ->
+      Error
+  end.
+
+%% @private
+-spec make_pending_acks(offset(), offset()) -> pending_acks().
+make_pending_acks(AckedOffset, BeginOffset) when
+    AckedOffset >= BeginOffset - 1 ->
+  #pending_acks{};
+make_pending_acks(AckedOffset, BeginOffset) ->
+  Begin = AckedOffset + 1,
+  End = BeginOffset - 1,
+  Count = End - Begin + 1,
+  Queue = queue:in({Begin, End}, queue:new()),
+  #pending_acks{offsets_queue = Queue, count = Count}.
```

Comment:
Kafka offsets may not be sequential, there can be gaps in compacted topics-partitions.
e.g. if message 1 has key 'a' and message 2, 3 both has key 'b', eventually message 2 will be compacted away leaving only message 1 and 3 on disk.

If we use `begin - acked` as the initial pre-fetch count,
`brod_consumer` may end up stuck with a too large made-up pending queue
and not fetching anything until the last message is acked.

e.g. if a message set has two messages `[offset = 1, offset = 99]`,
and if `brod_consumer` having a `prefetch_count = 10`,
it will not fetch anything until `offset = 99` is acked. 

To be accurate, the same offset segment queue (`queue:queue({Bgn::offset(), End::offset()})`)
should be tracked in `brod_xxx_subscriber`, and give this queue back to `brod_consumer` when re-subscribe.

I'm fine keeping current implementation.
I'd say this is a minor defect as 'fetch too few', which is better than 'fetch too many',
however it justifies a comment here when making up the initial pending queue.